### PR TITLE
feat(enhancement): Make ships snap to formation

### LINKED ISSUE #10285: Formation flying: Less corrections for staying in formation
### Problem Description

Ships move a bit restlessly when flying in formation. They perform corrections to get to their positions often, and this often causes holes in the formation.

### Related Issue Links

N/A

### Desired Solution

Ships stay in formation a little bit better, without as much corrections as is done now. Station-keeping seems much more stable. The expectation is that formation flying should be able to also be a lot more stable.

### Alternative Approaches

There have been some discussions if formations should do some "cheating" by automatically getting a little bit of backwards force on ships, or if the ships should be able to maintain position without any "cheating".

### Additional Context

Position stability didn't get a lot of attention for formation flying yet, so it seems like a good idea to look into this a bit more.

### PR DESCRIPTION
**Pull request by tibetiroka**
**Feature**

This PR addresses the bug/feature described in issue #10285 (fixes #10285)

## Summary
Adds automatic snapping to the position and velocity of ships once they are in formation. The position is adjusted continuously over multiple frames to make it appear smooth.

## Testing Done
I tested that ships require far less corrections to stay in formation (none if the lead ship doesn't accelerate), and that major course changes by the lead ship still require the ships to manually move back to position.

## Performance Impact
Should be minimal.

### CODE REVIEW

File: source/Ship.h [hunk 1]
```diff
@@ -300,7 +301,7 @@ class Ship : public Body, public std::enable_shared_from_this<Ship> {
 	const std::vector<EnginePoint> &ReverseEnginePoints() const;
 	const std::vector<EnginePoint> &SteeringEnginePoints() const;
 
-	// Make a ship disabled or destroyed, or bring back a destroyed ship.
+	// Make a ship disabled or destroyed, orVelocity bring back a destroyed ship.
```
Comment (RisingLeaf) [inline source/Ship.h:304]: huh?

Comment (tibetiroka) [reply source/Ship.h]: Oops.

**Comment by petervdmeer** [comment]
I've tested this with the `"line (sides)"` formation. The ships indeed stay in position very nicely:
line_sides_test1

But when I turn the flagship around (from stand-still), then the 2 ships directly besides the flagship (on both sides) just magically move to their new positions, without ever needing to change the orientation away from the orientation of the flagship, and without using thrusters anywhere in the process:
line_sides_test2

It looks a bit unnatural.

**Comment by petervdmeer** [comment]
Most formations will be small (max 7 ships), so most ships would be in the region close to the flagship.

**Comment by tibetiroka** [comment]
Part of that is by design. Since ships don't try to reposition themselves in the same frame, turning the flagship would always cause them to lose synchronicity (which is one of the main points of formation flying). My method avoids this problem for small corrections. It does look somewhat unnatural, but it also looks kinda nice if you forget about the physics.

Luckily, we can tweak this behaviour quite easily. I could lower the amount of snapping per frame, which would make ships less likely to stick to the formation. Alternatively, I could remove position snapping and only match velocities, which still solves the issue with ships repositioning themselves when the flagship doesn't change direction, but doesn't make the formations look that good (because the ships would still be slightly misaligned).

Which option do you like the most?

**Comment by petervdmeer** [comment]
What I would like to see is that ships stay in formation without making course-corrections when the flagship is not making any course-corrections (including when the flagship is standing still). This PR accomplishes that. Matching velocities _over time_ seems to be required for this.

But what I also would like to see is that ships do something like course-corrections to get to their positions when the flagship is making course-corrections. Just automatically sticking to their positions seems just wrong, as if we are looking at a clock, not at a group of spaceships that coordinate their behaviour.

The imperfections on positions give a little bit of a "human pilots" vibe, but staying in position perfectly looks better in the end, so I wouldn't mind if we also synchronize to perfect position-correctness _over time_ (besides having exactly matching velocities). We might also have military personalities where the formations match exactly, and merchant personalities where the formations are a little bit off.

If we would have [states for ships-AI](https://github.com/endless-sky/endless-sky/issues/10042), then I can imagine that we would implement the velocity/position matching with states like:
- FORMATION_MOVETO: Ship moves to the relative position (using the current code).
- FORMATION_VELOCITY: The the ship turns directly against the relative delta-velocity-vector (and 30% side wards on top of that towards the desired point, if the desired point is not directly on the delta-velocity vector), and makes a burn once it behind the desired point.
  - In this state, the velocity and position doesn't get adapted as a regular thruster burn, but brings the velocity to match exactly with the flagship and moves the ship a bit towards the desired point.
- FORMATION_POSITION: The ship turns directly towards the position where it should be (if is is not there already), and makes a burn once it is facing this point exactly.
  - In this state, the velocity doesn't get adapted at all, but causes the position delta to go to zero.
- FORMATION_KEEP: The ship just faces the same direction as the flagship, while velocity and position match exactly.
  - This is the formations idle state. Ship moves to state FORMATION_MOVETO if the position or velocity becomes too far off.

The nice thing about this implementation is that every course-correction is explicit (although there is an implicit drag in the FORMATION_POSITION state, but I would hope that to go unnoticed). It might be possible to implement all of this without state-machines, but that seems a lot more complex, so I filed the feature-request for ship-state-machines before trying to fix the last formation-movement-issues.

**Comment by petervdmeer** [comment]
I also experimented with another algorithm earlier:
- If a ships velocity is a little bit too much leftwards, and it is not too far right of its desired position, then slowly take out the leftward component of the delta-movement-vector.
- If a ships velocity is a little bit too much rightwards, and it is not too far left of its desired position, then slowly take out the rightward component of the delta-movement-vector.
- If a ships velocity is a little bit too fast forward, and it is not too far behind its desired position, then slowly take out the forward component of the delta-movement-vector.
- If a ships velocity is a little bit too slow (relative backwards movement vector), then don't correct for that. The ship can easily fix this by itself with thrusters and minimal movement. (And we don't want to give unfair speedboosts, only braking.)
- If the velocity is matching exactly, and the position is not too far off, then slowly move the ship to its desired position.

Never got to the point of actually testing and tuning this though (and I'm also not sure if this algorithm would look natural). The main consideration that I had for the slow corrections is that the player should not see it. (But it should be sufficient to avoid the continuous active course-corrections.)

**Comment by tibetiroka** [comment]
I removed position snapping entirely, and made velocity snapping a lot smoother. It still avoids idle position adjustments, but no longer keeps ships in formation when the flagship moves.

**Review by petervdmeer — APPROVED** [review]
Retested; looks good now.

**Pull request merged**